cmd/env.go

```diff
@@ -53,24 +86,113 @@ func env(cmd *cobra.Command, args []string) error {
-// is a string that can safely be used as one token in a shell command line.
-func shellescape(s string) string {
-	if len(s) == 0 {
-		return "''"
+// output will be returned lexically sorted by key name
+func buildEnvOutput(params map[string]string) ([]string, error) {
+	out := []string{}
+	for _, key := range sortedKeys(params) {
+		name := sanitizeKey(key)
+		if !preserveCase {
+			name = strings.ToUpper(name)
+		}
+
+		if err := validateShellName(name); err != nil {
+			return nil, err
+		}
+
+		// the default format prints all escape sequences as
+		// string literals, and wraps values in single quotes
+		// if they're unsafe or multi-line strings.
+		s := fmt.Sprintf(`%s=%s`, name, shellescape.Quote(params[key]))
+		if escapeSpecials {
+			// this format collapses special characters like newlines
+			// or carriage returns. requires escape sequences to be interpolated
+			// by whatever parses our key="value" pairs.
+			s = fmt.Sprintf(`%s="%s"`, name, doubleQuoteEscape(params[key]))
+		}
+
+		// don't rely on printf to handle properly quoting or
+		// escaping shell output -- just white-knuckle it ourselves.
+
+		out = append(out, s)
 	}
-	if pattern.MatchString(s) {
-		return "'" + strings.Replace(s, "'", "'\"'\"'", -1) + "'"
+
+	return out, nil
+}
+
+// The name of a variable can contain only letters (a-z, case insensitive),
+// numbers (0-9) or the underscore character (_). It may only begin with
+// a letter or an underscore.
+func validateShellName(s string) error {
+	shellChars := regexp.MustCompile(`^[A-Za-z0-9_]+$`).MatchString
+	validShellName := regexp.MustCompile(`^[A-Za-z_]{1}`).MatchString
+
+	if !shellChars(s) {
+		return fmt.Errorf("cmd: %q contains invalid characters for a shell variable name", s)
+	}
+
+	if !validShellName(s) {
+		return fmt.Errorf("cmd: shell variable name %q must start with a letter or underscore", s)
 	}
 
+	return nil
+}
+
+// note that all character width will be preserved; a single space
+// (or period, tab, or newline) will be replaced with a single underscore.
+// no squeezing/collapsing of replaced characters is performed at all.
+func sanitizeKey(s string) string {
+	// I promise, we don't actually care about allocations here.
+	// allocate *away*.
+	s = strings.TrimSpace(s)
+	s = strings.ReplaceAll(s, "-", "_")
+	s = strings.ReplaceAll(s, ".", "_")
+	// whitespace gets a visit from The Big Hammer that is regex.
+	s = regexp.MustCompile(`[[:space:]]`).ReplaceAllString(s, "_")
+
 	return s
 }
+
+// originally ported from github.com/joho/godotenv
+func doubleQuoteEscape(line string) string {
+	for _, c := range doubleQuoteSpecialChars {
+		toReplace := "\\" + string(c)
+		if c == '\n' {
+			toReplace = `\n`
+		}
+		if c == '\r' {
+			toReplace = `\r`
+		}
+		line = strings.Replace(line, string(c), toReplace, -1)
+	}
+	return line
+}
+
+func sortedKeys(params map[string]string) []string {
+	keys := make([]string, len(params))
+	i := 0
```

Comment:
What is this doing? I just see it get incremented down below but not actually used anywhere

Reply:
check again -- line 193. it's used as the index to assign values deterministically to `keys`. that being said, this function was moved here wholesale from export.go, with no changes. it's probably better expressed as a 3-part loop.

Reply:
🤦 
I read the loop body like 3 times and missed it.

Thanks.